Fixed part of #4057: FormBuilder tests

unit-00 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include 
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue 
  - when this PR is merged.
  -->
Fixes part of #4057:
Hi hello, this is my first time writing tests. This test is for FormBuilder, there were already a FormBuilderSpec.js so I added to it in an attempt to reach 100% coverage.

Should I added tests for `applyValidation` directive as well?

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [ ] The PR explanation includes the words "Fixes #bugnum: ...".
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

unit-00 committed:
Added additional valid unicode test for convertUnicodeWithParamsToHtml

unit-00 committed:
Added additional valid unicode test for convertUnicodeWithParamsToHtml

unit-00 committed:
Merge branch 'develop' of github.com:oppia/oppia into develop

unit-00 committed:
Added tests for valid unicode strings and requireIsFloat directive from FormBuilder.js

unit-00 committed:
Added tests for valid unicode strings and requireIsFloat directive from FormBuilder.js

unit-00 committed:
Fixed a semicolon to satisfy lint

unit-00 committed:
Fixed a semicolon to satisfy lint

unit-00 committed:
Merge branch 'issue-4057' of github.com:unit-00/oppia into issue-4057

seanlip commented:
Hi @unit-00, that's great, thanks for your first PR to Oppia!

 One question -- what is the current coverage percentage for this file (after your tests have been written)? Are the tests needed for applyValidation the only thing that's missing to get 100% coverage? If so I would suggest adding those here, too.

seanlip commented:
Also, btw, I just wanted to say that the tests that exist so far look great :)

unit-00 commented:
@seanlip Ahh! Hi hello Sean, the current coverage for this file is at 82.76% The tests needed for applyValidation is, I believe, the only thing that's needed to achieve close to 100% coverage. There is one small bit of code that's from `isFloat` filter in the FormBuilder that I have not been able to trigger an error. 
```    
try {
      var viewValue = input.toString().trim();
    } catch (e) {
      return undefined;
    }
```
I tried testing a wide range of values for input, however, they all turn into String in some way, even Error itself. I'll be sure to look into how I can trigger an error to be caught. 
As for `applyValidation`, I will be sure to do so.

Thank you so much for your review.

seanlip commented:
You know, I can't think of a way for .toString() to throw an error, either! I think we should just remove that try/catch there, it's pretty useless :)

(Btw, when writing tests, it's fine to modify the code under test if you think you can make it better. Tests usually expose lots of the weaknesses of the code so we might as well take advantage of that and fix them on the spot!)

unit-00 committed:
Added undefine case for normalizer and also for requireIsFloat

unit-00 committed:
added tests for applyValidation directive

unit-00 committed:
Merge branch 'issue-4057' into HEAD

unit-00 committed:
Merge branch 'develop' of github.com:oppia/oppia into HEAD

unit-00 committed:
fixed lint problems

unit-00 commented:
Hi hello @seanlip , I hope I haven't messed anything up. 

I've updated the file to address the tests for applyValidation. Currently, the coverage for FormBuilder.js went from 82.76% to 98.85%, leaving only line 195 of FormBulider.js. I was able to manage hit an error with the previous `input.toString().trim()` by using `undefined` for input, though I am afraid of it being a trivial case. 

Please review over the updated files again and tell me if there's anything else I need to do. Thank you for your time.

seanlip commented:
Hi @unit-00, I think I mentioned removing the try/catch in a previous comment. How about doing that?

seanlip commented on core/templates/dev/head/components/forms/FormBuilderSpec.js:
```diff
@@ -146,6 +169,212 @@ describe('Normalizer tests', function() {
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue(-3);
```
In both this and the next test I think it makes sense to add a value that lies on the boundary (e.g. -2.5 in this case and 5 in the next)?

seanlip commented on core/templates/dev/head/components/forms/FormBuilderSpec.js:
```diff
@@ -146,6 +169,212 @@ describe('Normalizer tests', function() {
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue(undefined);
```
This one is probably not going to happen in practice, I think? (If that's the case, no harm adding it here, but also probably not necessary IMO.)

seanlip commented on core/templates/dev/head/components/forms/FormBuilderSpec.js:
```diff
@@ -146,6 +169,212 @@ describe('Normalizer tests', function() {
+    expect(testInput.$valid).toEqual(false);
+  }));
+
+  it('should apply isFloat validation', inject(function($compile) {
```
Interesting. Is the result ever false? Maybe add a test that uses ints?

unit-00 replied:
I tried using strings, e.g. 'abc' to get a false case, but the tests cough up an undefined instead. I wasn't sure how to get a false case.  I'll definitely try again.

seanlip replied:
Hm, I took a look at the original method and I can't see how it returns false either. It seems a bit weird actually -- whereas all the other isInteger, isNonempty, etc. methods return true or false, this one seems to return a float or undefined. That seems inconsistent. But it's great we're catching this -- that's what happens when you write tests :)

I suggest changing the main code to return true/false instead, and updating the tests accordingly. Just make sure to take a look at all instances of isFloat  -- e.g. in SchemaBasedFloatEditorDirective.js -- and verify that everything still works correctly. Also test manually with a float editor (e.g. in the NumericInput interaction perhaps?) to see whether everything still behaves OK when you type in valid input and when you do not.

If there are issues, then maybe what's really needed is two filters -- isFloat, and parseAsFloat. But let's see if we can do without the second one.

Does that sound good?

seanlip reviewed (changes requested):
Hi @unit-00, I think this is great, thanks! I like that you've covered all the cases pretty thoroughly, too!

Left some inline comments, PTAL.

unit-00 commented:
Hey @seanlip ! I think those are two different try/catch, one from `isFloat` filter and the other one from `applyValidation` directive. I was able to trigger the error from the first try/catch, the `toString().trim()`, via undefined. Should I remove the both of them?

 I believe you're right regarding the undefined cases as I don't think they'll happen in practice either, I'll be sure to remove that case. Along with that, I'll fix up the rest of cases that you mentioned!

unit-00 committed:
Merge remote-tracking branch 'upstream/develop' into issue-4057

unit-00 committed:
Added boundary cases, took out impractical cases, and added an additional test case to isFloat

unit-00 committed:
Changed the undefined cases in isFloat to errors not equal 0

seanlip commented:
Hmm. On second thoughts, I think I would keep both, actually. I can imagine undefined potentially occurring when the input field is initially loading perhaps, and it doesn't really do any harm to keep it there I think.

You can probably trigger the second one by using the name of a nonexistent filter perhaps?

unit-00 committed:
removed a try/catch block in isFloat filter

unit-00 committed:
removed test cases where we test for undefined inputs

unit-00 committed:
rewrote error test cases in applyValidation to be semantically understandable

unit-00 committed:
Merge remote-tracking branch 'upstream/develop' into issue-4057

unit-00 committed:
Adding try/catch block back

unit-00 committed:
Added undefined cases back

unit-00 committed:
Added nonexistent filter case

unit-00 committed:
fixed lint issues

unit-00 commented:
Hi hello @seanlip , I have done it! I've fixed all the things you've commented on earlier and also added a case for a nonexistent filter to be applied. Coverage currently is at 100%. 

I accidentally took out the try/catch before reading the comments, I've added them back. hahaha

seanlip commented on core/templates/dev/head/components/forms/FormBuilderSpec.js:
```diff
@@ -146,6 +169,250 @@ describe('Normalizer tests', function() {
+
+    testInput.$setViewValue('O');
+    scope.$digest();
+    expect(Object.keys(testInput.$error).length).not.toEqual(0);
```
I think if you're including tests with checks on both testInput.$valid and Object.keys(testInput.$error).length, you should check _both_ of these after _each_ change of the viewValue, rather than check one condition sometimes and the other condition otherwise.

Ditto below.

unit-00 replied:
Ahh ahh! I believe you're right, as it'll bring thoroughness. I'll be sure to add them. 

Should I add them to all the filters or just `isFloat` and `isInteger`, as they're the only tests where both of these appear?

Edit: I've added both checks, $valid and the length of the $errors, to all filters for thoroughness for now. If you'd like me to take them out, I'll do so.

seanlip reviewed (commented):
Hi @unit-00, it looks great, thanks! Just one more comment and then LGTM.

unit-00 committed:
Added tests for error for thoroughness

unit-00 committed:
Merge remote-tracking branch 'upstream/develop' into issue-4057

unit-00 commented:
Hey @seanlip ! My previous comments disappeared. I've added both checks, $valid and the length of $error, to all the filters for thoroughness. 

If you'd like to take it out and just leave them in just for `isFloat` and `isInteger`, please tell me so!

seanlip reviewed (approved):
I think it's all good! Thanks @unit-00, and congrats on your first PR to Oppia :)
